l10n_fi_invoice/models/account_move.py

```diff
@@ -0,0 +1,85 @@
+##############################################################################
+#
+#    Author: Avoin.Systems
+#    Copyright 2015-2019 Avoin.Systems
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from odoo import models, fields
+# noinspection PyProtectedMember
+from odoo.tools.translate import _
+import re
+import logging
+
+log = logging.getLogger(__name__)
+
+
+class AccountMove(models.Model):
+    _inherit = 'account.move'
+
+    def _get_invoice_pdf_filename(self):
+        type_string = 'Invoice'
+        invoice_numbers = self.name or ''
+        if self.type in ('in_refund', 'out_refund'):
+            type_string = 'Refund'
+            invoice_numbers = '-'.join((
+                invoice_numbers, self.refund_invoice_id.name))
+        filename = '-'.join((
+            type_string,
+            invoice_numbers,
+            self.company_id.display_name,
+            self.partner_id.display_name)). \
+            replace(' ', '-').replace(',', '').replace('--', '-')
+        return filename
+
+    def _compute_barcode_string(self):
+        for invoice in self:
+            displayed_bank_accounts = invoice.company_id.partner_id.bank_ids.filtered('journal_id.include_on_invoice')
+            primary_bank_account = invoice.invoice_partner_bank_id or \
+                displayed_bank_accounts and displayed_bank_accounts[0]
+            if (invoice.amount_total and primary_bank_account.acc_number
+                    and invoice.ref_number and invoice.invoice_date_due):
+                amount_total_string = str(invoice.amount_total)
```

Comment:
We should make some extra checks for correctness:
1. We should use the currency to round the `amount_total` so we don't get float-weirdness
2. We should let Python's `format` ensure we always get two decimal places.
3. We should `split` the string into euros and cents for clarity.
4. We should handle the case where the amount of the invoice is too large (greater than 6 digits, i.e. in the millions)

```python
BARCODE_AMOUNT_LIMIT = 999999.99
...
if invoice.currency_id.compare(BARCODE_AMOUNT_LIMIT, invoice.amount_total) == 1:
    # Amount is too large, we choose the option of replacing the amount with zeros (in accordance with  
    # https://www.finanssiala.fi/maksujenvalitys/dokumentit/Pankkiviivakoodi-opas.pdf)
    # so that at least some use can be gotten from the barcode
    euros = "000000"
    cents = "00"
else:
    full_amount = invoice.currency_id.round(invoice.amount_total)
    euros, cents = "{:.2f}".format(full_amount).zfill(9).split('.')
...
```

Reply:
I implemented these additional checks.